packages/resolve-dependencies/src/index.ts

```diff
@@ -183,8 +183,13 @@ export default async function (
 
   if (opts.forceFullResolution && opts.wantedLockfile != null) {
     for (const [depPath, pkg] of Object.entries(dependenciesGraph)) {
-      // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
-      if (opts.neverBuiltDependencies?.has(pkg.name) || opts.wantedLockfile.packages?.[depPath] == null || pkg.requiresBuild) continue
+      if (
+        // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+        opts.neverBuiltDependencies?.has(pkg.name) ||
+        (opts.onlyBuiltDependencies !== false && opts.onlyBuiltDependencies.has(pkg.name)) ||
+        opts.wantedLockfile.packages?.[depPath] == null ||
+        pkg.requiresBuild
+      ) continue
       pendingRequiresBuilds.push(depPath)
```

Comment:
`opts.onlyBuiltDependencies !== false && opts.onlyBuiltDependencies.has(pkg.name)` then `continue` instead of adding the `pkg` to `pendingRequiresBuilds`? No wonder why the test failed.